precompile/warp_messenger.go

```diff
@@ -0,0 +1,425 @@
+		return nil, remainingGas, err
+	}
+
+	// Return the packed output and the remaining gas
+	return packedOutput, remainingGas, nil
+}
+
+// UnpackGetVerifiedWarpMessageInput attempts to unpack [input] into the *big.Int type argument
+// assumes that [input] does not include selector (omits first 4 func signature bytes)
+func UnpackGetVerifiedWarpMessageInput(input []byte) (*big.Int, error) {
+	res, err := WarpMessengerABI.UnpackInput("getVerifiedWarpMessage", input)
+	if err != nil {
+		return big.NewInt(0), err
+	}
+	unpacked := *abi.ConvertType(res[0], new(*big.Int)).(**big.Int)
+	return unpacked, nil
+}
+
+// PackGetVerifiedWarpMessage packs [messageIndex] of type *big.Int into the appropriate arguments for getVerifiedWarpMessage.
+// the packed bytes include selector (first 4 func signature bytes).
+// This function is mostly used for tests.
+func PackGetVerifiedWarpMessage(messageIndex *big.Int) ([]byte, error) {
+	return WarpMessengerABI.Pack("getVerifiedWarpMessage", messageIndex)
+}
+
+// PackGetVerifiedWarpMessageOutput attempts to pack given [outputStruct] of type GetVerifiedWarpMessageOutput
+// to conform the ABI outputs.
+func PackGetVerifiedWarpMessageOutput(outputStruct GetVerifiedWarpMessageOutput) ([]byte, error) {
+	return WarpMessengerABI.PackOutput("getVerifiedWarpMessage",
+		outputStruct.Message,
+		outputStruct.Success,
+	)
+}
+
+func getVerifiedWarpMessage(accessibleState PrecompileAccessibleState, caller common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+	if remainingGas, err = deductGas(suppliedGas, GetVerifiedWarpMessageGasCost); err != nil {
+		return nil, 0, err
+	}
+	if readOnly {
+		return nil, remainingGas, vmerrs.ErrWriteProtection
+	}
+	// attempts to unpack [input] into the arguments to the GetVerifiedWarpMessageInput.
+	// Assumes that [input] does not include selector
+	// You can use unpacked [messageIndex] variable in your code
+	inputIndex, err := UnpackGetVerifiedWarpMessageInput(input)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	storageSlots, exists := accessibleState.GetStateDB().GetPredicateStorageSlots(WarpMessengerAddress)
+	if !exists || storageSlots == nil {
+		return nil, remainingGas, ErrMissingStorageSlots
+	}
+
+	var signedMessages [][]byte
+	err = rlp.DecodeBytes(storageSlots, &signedMessages)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Check that the message index exists.
+	if !inputIndex.IsInt64() {
+		return nil, remainingGas, ErrInvalidMessageIndex
+	}
+	messageIndex := inputIndex.Int64()
+	if len(signedMessages) <= int(messageIndex) {
+		return nil, remainingGas, ErrInvalidMessageIndex
+	}
+
+	// Parse the raw message to be processed.
+	signedMessage := signedMessages[messageIndex]
+	message, err := teleporter.ParseMessage(signedMessage)
+	if err != nil {
+		return nil, remainingGas, err
+	}
+
+	// Charge gas per validator included in the aggregate signature
```

Comment:
@aaronbuchwald correct me if I'm wrong here, but we'll likely want to remove the gas charge per signature aggregated from the precompile. Instead, Aaron mentioned handling it along with the intrinsic gas of the transaction. The public key aggregation and signature verification occurs whether or not the transaction actually calls `getVerifiedWarpMessage`, so charging here doesn't prevent that DOS vector.